expression/builtin_other.go

```diff
@@ -103,14 +134,59 @@ func (c *inFunctionClass) getFunction(ctx sessionctx.Context, args []Expression)
 	return sig, nil
 }
 
+type baseInSig struct {
+	baseBuiltinFunc
+	nonConstArgs []Expression
+	// hashSet will not be used when the number of constant arguments is smaller than threshold
+	threshold int
+	hasNull   bool
+}
+
 // builtinInIntSig see https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_in
 type builtinInIntSig struct {
-	baseBuiltinFunc
+	baseInSig
+	hashSet map[int64]bool
+}
+
+func (b *builtinInIntSig) buildHashMapForConstArgs(ctx sessionctx.Context) error {
+	b.nonConstArgs = make([]Expression, 0, len(b.args))
+	b.nonConstArgs = append(b.nonConstArgs, b.args[0])
```

Comment:
I think the assumption that almost all the argument is constant is more reasonable, which result in the following initialization codes:
```go
b.nonConstArgs := []Expression{b.args[0]}
b.hashSet := make(map[int64]bool, len(b.args)-1)
```